File: web/src/__tests__/dsl/directives/DirectiveSegmenter.test.ts
import { describe, it, expect } from 'vitest'
import { segmentLine, buildDisplayText, hasDirectives, isViewSegment, type DirectiveSegmentType } from '../../../dsl/directives/DirectiveSegmenter'
import type { DirectiveResult } from '../../../dsl/directives/DirectiveResult'
import { directiveResultSuccess, directiveResultFailure } from '../../../dsl/directives/DirectiveResult'
import { directiveHash } from '../../../dsl/directives/DirectiveFinder'
import { numberVal, stringVal, alarmVal, viewVal } from '../../../dsl/runtime/DslValue'
import type { Note } from '../../../data/Note'

function resultsMap(entries: [string, DirectiveResult][]): Map<string, DirectiveResult> {
  return new Map(entries)
}

describe('segmentLine', () => {
  it('returns single text segment for plain text', () => {
    const segments = segmentLine('hello world', 'test-line', new Map())
    expect(segments).toHaveLength(1)
    expect(segments[0]!.kind).toBe('Text')
    if (segments[0]!.kind === 'Text') {
      expect(segments[0]!.content).toBe('hello world')
      expect(segments[0]!.rangeStart).toBe(0)
      expect(segments[0]!.rangeEnd).toBe(11)
    }
  })

  it('returns empty array for empty content', () => {
    const segments = segmentLine('', 'test-line', new Map())
    expect(segments).toHaveLength(0)
  })

  it('returns directive segment for directive-only content', () => {
    const segments = segmentLine('[test]', 'test-line', new Map())
    expect(segments).toHaveLength(1)
    expect(segments[0]!.kind).toBe('Directive')
    if (segments[0]!.kind === 'Directive') {
      expect(segments[0]!.sourceText).toBe('[test]')
      expect(segments[0]!.key).toBe(directiveHash('[test]'))
      expect(segments[0]!.result).toBeNull()
      expect(segments[0]!.isComputed).toBe(false)
    }
  })

  it('splits text and directive segments', () => {
    const segments = segmentLine('before [dir] after', 'test-line', new Map())
    expect(segments).toHaveLength(3)
    expect(segments[0]!.kind).toBe('Text')
    expect(segments[1]!.kind).toBe('Directive')
    expect(segments[2]!.kind).toBe('Text')

    if (segments[0]!.kind === 'Text') expect(segments[0]!.content).toBe('before ')
    if (segments[1]!.kind === 'Directive') expect(segments[1]!.sourceText).toBe('[dir]')
    if (segments[2]!.kind === 'Text') expect(segments[2]!.content).toBe(' after')
  })

  it('uses directive hash for key', () => {
    const segments = segmentLine('[test]', 'note-abc', new Map())
    expect(segments).toHaveLength(1)
    if (segments[0]!.kind === 'Directive') {
      expect(segments[0]!.key).toBe(directiveHash('[test]'))
    }
  })

  it('attaches result when available', () => {
    const key = directiveHash('[calc]')
    const result = directiveResultSuccess(numberVal(42))
    const segments = segmentLine('[calc]', 'test-line', resultsMap([[key, result]]))
    expect(segments).toHaveLength(1)
    if (segments[0]!.kind === 'Directive') {
      expect(segments[0]!.result).not.toBeNull()
      expect(segments[0]!.isComputed).toBe(true)
      expect(segments[0]!.displayText).toBe('42')
    }
  })

  it('shows source text when no result', () => {
    const segments = segmentLine('[calc]', 'test-line', new Map())
    expect(segments).toHaveLength(1)
    if (segments[0]!.kind === 'Directive') {
      expect(segments[0]!.displayText).toBe('[calc]')
      expect(segments[0]!.isComputed).toBe(false)
    }
  })

  it('shows source text for error result', () => {
    const key = directiveHash('[calc]')
    const result = directiveResultFailure('bad')
    const segments = segmentLine('[calc]', 'test-line', resultsMap([[key, result]]))
    expect(segments).toHaveLength(1)
    if (segments[0]!.kind === 'Directive') {
      // isComputed is false for errors, so displayText remains sourceText
      expect(segments[0]!.isComputed).toBe(false)
      expect(segments[0]!.displayText).toBe('[calc]')
    }
  })

  it('handles multiple directives', () => {
    const segments = segmentLine('[a] + [b]', 'test-line', new Map())
    expect(segments).toHaveLength(3)
    expect(segments[0]!.kind).toBe('Directive')
    expect(segments[1]!.kind).toBe('Text')
    expect(segments[2]!.kind).toBe('Directive')
  })

  it('tracks correct ranges for segments', () => {
    const segments = segmentLine('hi [x] bye', 'test-line', new Map())
    expect(segments).toHaveLength(3)

    // "hi " -> 0..3
    if (segments[0]!.kind === 'Text') {
      expect(segments[0]!.rangeStart).toBe(0)
      expect(segments[0]!.rangeEnd).toBe(3)
    }
    // "[x]" -> 3..6
    if (segments[1]!.kind === 'Directive') {
      expect(segments[1]!.rangeStart).toBe(3)
      expect(segments[1]!.rangeEnd).toBe(6)
    }
    // " bye" -> 6..10
    if (segments[2]!.kind === 'Text') {
      expect(segments[2]!.rangeStart).toBe(6)
      expect(segments[2]!.rangeEnd).toBe(10)
    }
  })
})

describe('buildDisplayText', () => {
  it('returns plain text unchanged', () => {
    const result = buildDisplayText('hello', 'test-line', new Map())
    expect(result.displayText).toBe('hello')
    expect(result.directiveDisplayRanges).toHaveLength(0)
  })

  it('returns empty for empty content', () => {
    const result = buildDisplayText('', 'test-line', new Map())
    expect(result.displayText).toBe('')
    expect(result.segments).toHaveLength(0)
  })

  it('replaces directive source with computed result', () => {
    const key = directiveHash('[calc]')
    const dr = directiveResultSuccess(numberVal(42))
    const result = buildDisplayText('[calc]', 'test-line', resultsMap([[key, dr]]))
    expect(result.displayText).toBe('42')
  })

  it('keeps directive source when no result', () => {
    const result = buildDisplayText('val=[calc]', 'test-line', new Map())
    expect(result.displayText).toBe('val=[calc]')
  })

  it('builds display text with mixed text and directives', () => {
    const key = directiveHash('[greet]')
    const dr = directiveResultSuccess(stringVal('world'))
    const result = buildDisplayText('say [greet]!', 'test-line', resultsMap([[key, dr]]))
    expect(result.displayText).toBe('say world!')
  })

  it('populates directive display ranges', () => {
    const key = directiveHash('[x]')
    const dr = directiveResultSuccess(numberVal(7))
    const result = buildDisplayText('[x] end', 'test-line', resultsMap([[key, dr]]))
    expect(result.directiveDisplayRanges).toHaveLength(1)

    const range = result.directiveDisplayRanges[0]!
    expect(range.key).toBe(key)
    expect(range.sourceText).toBe('[x]')
    expect(range.displayText).toBe('7')
    expect(range.isComputed).toBe(true)
    expect(range.hasError).toBe(false)
    expect(range.hasWarning).toBe(false)
    expect(range.sourceRangeStart).toBe(0)
    expect(range.sourceRangeEnd).toBe(3)
    expect(range.displayRangeStart).toBe(0)
    expect(range.displayRangeEnd).toBe(1) // "7" is 1 char
  })

  it('marks error in display range', () => {
    const key = directiveHash('[x]')
    const dr = directiveResultFailure('bad')
    const result = buildDisplayText('[x]', 'test-line', resultsMap([[key, dr]]))
    expect(result.directiveDisplayRanges).toHaveLength(1)
    expect(result.directiveDisplayRanges[0]!.hasError).toBe(true)
    expect(result.directiveDisplayRanges[0]!.isComputed).toBe(false)
  })

  it('handles multiple directives with different display lengths', () => {
    const key1 = directiveHash('[xxx]')
    const key2 = directiveHash('[yyy]')
    const dr1 = directiveResultSuccess(stringVal('AB'))
    const dr2 = directiveResultSuccess(stringVal('CDEF'))
    const result = buildDisplayText('[xxx] [yyy]', 'test-line', resultsMap([[key1, dr1], [key2, dr2]]))
    expect(result.displayText).toBe('AB CDEF')

    const ranges = result.directiveDisplayRanges
    expect(ranges).toHaveLength(2)
    // First: "AB" at display 0..2
    expect(ranges[0]!.displayRangeStart).toBe(0)
    expect(ranges[0]!.displayRangeEnd).toBe(2)
    // Second: "CDEF" at display 3..7 (after "AB ")
    expect(ranges[1]!.displayRangeStart).toBe(3)
    expect(ranges[1]!.displayRangeEnd).toBe(7)
  })
})

describe('buildDisplayText alarm directives', () => {
  it('marks alarm directive with isAlarm and alarmId', () => {
    const key = directiveHash('[alarm("alarm-123")]')
    const dr = directiveResultSuccess(alarmVal('alarm-123'))
    const result = buildDisplayText('[alarm("alarm-123")]', 'test-line', resultsMap([[key, dr]]))

    expect(result.directiveDisplayRanges).toHaveLength(1)
    const range = result.directiveDisplayRanges[0]!
    expect(range.isAlarm).toBe(true)
    expect(range.alarmId).toBe('alarm-123')
    expect(range.isButton).toBe(false)
    expect(range.isView).toBe(false)
  })

  it('non-alarm directive has isAlarm false and no alarmId', () => {
    const key = directiveHash('[calc]')
    const dr = directiveResultSuccess(numberVal(42))
    const result = buildDisplayText('[calc]', 'test-line', resultsMap([[key, dr]]))

    const range = result.directiveDisplayRanges[0]!
    expect(range.isAlarm).toBe(false)
    expect(range.alarmId).toBeUndefined()
  })

  it('alarm directive mixed with text preserves display ranges', () => {
    const key = directiveHash('[alarm("a1")]')
    const dr = directiveResultSuccess(alarmVal('a1'))
    const result = buildDisplayText('Task [alarm("a1")] done', 'test-line', resultsMap([[key, dr]]))

    expect(result.directiveDisplayRanges).toHaveLength(1)
    const range = result.directiveDisplayRanges[0]!
    expect(range.isAlarm).toBe(true)
    expect(range.alarmId).toBe('a1')
  })

  it('alarm directive displays as clock emoji', () => {
    const key = directiveHash('[alarm("a1")]')
    const dr = directiveResultSuccess(alarmVal('a1'))
    const result = buildDisplayText('[alarm("a1")]', 'test-line', resultsMap([[key, dr]]))
    expect(result.displayText).toBe('\u23F0')
  })

  it('alarm sourceRange covers entire directive text', () => {
    const key = directiveHash('[alarm("a1")]')
    const dr = directiveResultSuccess(alarmVal('a1'))
    const result = buildDisplayText('Task [alarm("a1")] done', 'test-line', resultsMap([[key, dr]]))

    const range = result.directiveDisplayRanges[0]!
    expect(range.sourceRangeStart).toBe(5)
    expect(range.sourceRangeEnd).toBe(18) // end of [alarm("a1")]
  })

  it('alarm displayRange is single emoji width', () => {
    const key = directiveHash('[alarm("a1")]')
    const dr = directiveResultSuccess(alarmVal('a1'))
    const result = buildDisplayText('[alarm("a1")]', 'test-line', resultsMap([[key, dr]]))

    const range = result.directiveDisplayRanges[0]!
    expect(range.displayRangeEnd - range.displayRangeStart).toBe(1) // "⏰" = 1 char
  })
})

describe('buildDisplayText recurringAlarm directives', () => {
  it('marks recurringAlarm directive with isAlarm and recurringAlarmId', () => {
    const key = directiveHash('[recurringAlarm("rec-123")]')
    const dr = directiveResultSuccess(alarmVal('rec-123'))
    const result = buildDisplayText('[recurringAlarm("rec-123")]', 'test-line', resultsMap([[key, dr]]))

    expect(result.directiveDisplayRanges).toHaveLength(1)
    const range = result.directiveDisplayRanges[0]!
    expect(range.isAlarm).toBe(true)
    expect(range.recurringAlarmId).toBe('rec-123')
    // alarmId is also set from the AlarmVal result value
    expect(range.alarmId).toBe('rec-123')
  })

  it('recurringAlarm directive displays as clock emoji', () => {
    const key = directiveHash('[recurringAlarm("rec1")]')
    const dr = directiveResultSuccess(alarmVal('rec1'))
    const result = buildDisplayText('[recurringAlarm("rec1")]', 'test-line', resultsMap([[key, dr]]))
    expect(result.displayText).toBe('\u23F0')
  })

  it('renders recurringAlarm as clock emoji even without computed result', () => {
    const result = buildDisplayText('[recurringAlarm("rec1")]', 'test-line', new Map())
    expect(result.displayText).toBe('\u23F0')
    expect(result.directiveDisplayRanges[0]!.isAlarm).toBe(true)
    expect(result.directiveDisplayRanges[0]!.recurringAlarmId).toBe('rec1')
  })

  it('mixed alarm and recurringAlarm on same line', () => {
    const key1 = directiveHash('[alarm("a1")]')
    const key2 = directiveHash('[recurringAlarm("rec1")]')
    const dr1 = directiveResultSuccess(alarmVal('a1'))
    const dr2 = directiveResultSuccess(alarmVal('rec1'))
    const result = buildDisplayText(
      '[alarm("a1")] [recurringAlarm("rec1")]', 'test-line',
      resultsMap([[key1, dr1], [key2, dr2]])
    )

    expect(result.directiveDisplayRanges).toHaveLength(2)
    expect(result.directiveDisplayRanges[0]!.alarmId).toBe('a1')
    expect(result.directiveDisplayRanges[0]!.recurringAlarmId).toBeUndefined()
    expect(result.directiveDisplayRanges[1]!.recurringAlarmId).toBe('rec1')
  })
})

function directiveSegment(sourceText: string, result: DirectiveResult | null): DirectiveSegmentType {
  return {
    kind: 'Directive',
    sourceText,
    key: directiveHash(sourceText),
    result,
    rangeStart: 0,
    rangeEnd: sourceText.length,
    displayText: sourceText,
    isComputed: false,
  }
}

const dummyNote: Note = {
  id: 'n1', userId: 'u1', parentNoteId: null, content: 'test',
  createdAt: null, updatedAt: null, lastAccessedAt: null,
  tags: [], containedNotes: [], state: null, path: '/test',
  rootNoteId: null, showCompleted: true,
}

describe('isViewSegment', () => {
  it('returns true for view directive with successful result', () => {
    const result = directiveResultSuccess(viewVal([dummyNote]))
    expect(isViewSegment(directiveSegment('[view(find())]', result))).toBe(true)
  })

  it('returns true for view directive with error result (source text fallback)', () => {
    const result = directiveResultFailure('parse error')
    expect(isViewSegment(directiveSegment('[view(bad syntax)]', result))).toBe(true)
  })

  it('returns true for view directive with no result', () => {
    expect(isViewSegment(directiveSegment('[view(find())]', null))).toBe(true)
  })

  it('returns false for non-view directive', () => {
    const result = directiveResultSuccess(numberVal(42))
    expect(isViewSegment(directiveSegment('[calc]', result))).toBe(false)
  })

  it('returns false for non-view directive with error', () => {
    const result = directiveResultFailure('bad')
    expect(isViewSegment(directiveSegment('[calc]', result))).toBe(false)
  })

  it('returns false for source text that merely contains view', () => {
    expect(isViewSegment(directiveSegment('[overview()]', null))).toBe(false)
  })
})

describe('buildDisplayText view directives', () => {
  it('marks view directive with isView when result is ViewVal', () => {
    const source = '[view(find())]'
    const key = directiveHash(source)
    const dr = directiveResultSuccess(viewVal([dummyNote]))
    const result = buildDisplayText(source, 'test-line', resultsMap([[key, dr]]))
    expect(result.directiveDisplayRanges[0]!.isView).toBe(true)
  })

  it('marks view directive with isView even on error (source text fallback)', () => {
    const source = '[view(bad)]'
    const key = directiveHash(source)
    const dr = directiveResultFailure('parse error')
    const result = buildDisplayText(source, 'test-line', resultsMap([[key, dr]]))
    expect(result.directiveDisplayRanges[0]!.isView).toBe(true)
  })

  it('marks view directive with isView when no result at all', () => {
    const source = '[view(find())]'
    const result = buildDisplayText(source, 'test-line', new Map())
    expect(result.directiveDisplayRanges[0]!.isView).toBe(true)
  })

  it('does not mark non-view directive as isView', () => {
    const source = '[calc]'
    const result = buildDisplayText(source, 'test-line', new Map())
    expect(result.directiveDisplayRanges[0]!.isView).toBe(false)
  })
})

describe('hasDirectives', () => {
  it('returns true for content with brackets', () => {
    expect(hasDirectives('[test]')).toBe(true)
    expect(hasDirectives('some [dir] here')).toBe(true)
  })

  it('returns false for plain text', () => {
    expect(hasDirectives('no directives')).toBe(false)
    expect(hasDirectives('')).toBe(false)
  })

  it('returns false for unclosed brackets', () => {
    expect(hasDirectives('[unclosed')).toBe(false)
  })
})

File: app/src/test/java/org/alkaline/taskbrain/dsl/directives/DirectiveSegmenterTest.kt
package org.alkaline.taskbrain.dsl.directives

import org.alkaline.taskbrain.dsl.runtime.NumberVal
import org.junit.Assert.*
import org.junit.Test

class DirectiveSegmenterTest {

    private companion object {
        const val TEST_LINE_ID = "test-line"
    }

    // region segmentLine tests

    @Test
    fun `segmentLine returns empty list for empty content`() {
        val segments = DirectiveSegmenter.segmentLine("", TEST_LINE_ID, emptyMap())
        assertTrue(segments.isEmpty())
    }

    @Test
    fun `segmentLine returns single text segment for content without directives`() {
        val segments = DirectiveSegmenter.segmentLine("Hello world", TEST_LINE_ID, emptyMap())

        assertEquals(1, segments.size)
        val segment = segments[0] as DirectiveSegment.Text
        assertEquals("Hello world", segment.content)
        assertEquals(0..10, segment.range)
    }

    @Test
    fun `segmentLine returns directive segment for single directive`() {
        val content = "[42]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val segments = DirectiveSegmenter.segmentLine(content, TEST_LINE_ID, results)

        assertEquals(1, segments.size)
        val segment = segments[0] as DirectiveSegment.Directive
        assertEquals("[42]", segment.sourceText)
        assertTrue(segment.isComputed)
        assertEquals("42", segment.displayText)
    }

    @Test
    fun `segmentLine handles text before and after directive`() {
        val content = "Hello [42] world"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val segments = DirectiveSegmenter.segmentLine(content, TEST_LINE_ID, results)

        assertEquals(3, segments.size)

        val text1 = segments[0] as DirectiveSegment.Text
        assertEquals("Hello ", text1.content)

        val directive = segments[1] as DirectiveSegment.Directive
        assertEquals("[42]", directive.sourceText)
        assertEquals("42", directive.displayText)

        val text2 = segments[2] as DirectiveSegment.Text
        assertEquals(" world", text2.content)
    }

    @Test
    fun `segmentLine handles multiple directives`() {
        val content = "[1] and [2] and [3]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val segments = DirectiveSegmenter.segmentLine(content, TEST_LINE_ID, results)

        assertEquals(5, segments.size)

        val d1 = segments[0] as DirectiveSegment.Directive
        assertEquals("[1]", d1.sourceText)

        val t1 = segments[1] as DirectiveSegment.Text
        assertEquals(" and ", t1.content)

        val d2 = segments[2] as DirectiveSegment.Directive
        assertEquals("[2]", d2.sourceText)

        val t2 = segments[3] as DirectiveSegment.Text
        assertEquals(" and ", t2.content)

        val d3 = segments[4] as DirectiveSegment.Directive
        assertEquals("[3]", d3.sourceText)
    }

    @Test
    fun `segmentLine shows source text when no result available`() {
        val content = "[42]"
        // No results provided
        val segments = DirectiveSegmenter.segmentLine(content, TEST_LINE_ID, emptyMap())

        assertEquals(1, segments.size)
        val segment = segments[0] as DirectiveSegment.Directive
        assertEquals("[42]", segment.sourceText)
        assertFalse(segment.isComputed)
        assertEquals("[42]", segment.displayText) // Source text when not computed
    }

    // endregion

    // region hasDirectives tests

    @Test
    fun `hasDirectives returns false for content without directives`() {
        assertFalse(DirectiveSegmenter.hasDirectives("Hello world"))
        assertFalse(DirectiveSegmenter.hasDirectives(""))
        assertFalse(DirectiveSegmenter.hasDirectives("No brackets here"))
    }

    @Test
    fun `hasDirectives returns true for content with directives`() {
        assertTrue(DirectiveSegmenter.hasDirectives("[42]"))
        assertTrue(DirectiveSegmenter.hasDirectives("Hello [42] world"))
        assertTrue(DirectiveSegmenter.hasDirectives("[\"string\"]"))
    }

    // endregion

    // region hasComputedDirectives tests

    @Test
    fun `hasComputedDirectives returns false for content without directives`() {
        assertFalse(DirectiveSegmenter.hasComputedDirectives("Hello world", TEST_LINE_ID, emptyMap()))
    }

    @Test
    fun `hasComputedDirectives returns false when no results available`() {
        assertFalse(DirectiveSegmenter.hasComputedDirectives("[42]", TEST_LINE_ID, emptyMap()))
    }

    @Test
    fun `hasComputedDirectives returns true when result is available`() {
        val content = "[42]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        assertTrue(DirectiveSegmenter.hasComputedDirectives(content, TEST_LINE_ID, results))
    }

    @Test
    fun `hasComputedDirectives returns false when result has error`() {
        val content = "[42]"
        val key = DirectiveResult.hashDirective("[42]")
        val errorResult = mapOf(key to DirectiveResult.failure("Error"))
        assertFalse(DirectiveSegmenter.hasComputedDirectives(content, TEST_LINE_ID, errorResult))
    }

    // endregion

    // region buildDisplayText tests

    @Test
    fun `buildDisplayText returns empty for empty content`() {
        val result = DirectiveSegmenter.buildDisplayText("", TEST_LINE_ID, emptyMap())
        assertEquals("", result.displayText)
        assertTrue(result.segments.isEmpty())
        assertTrue(result.directiveDisplayRanges.isEmpty())
    }

    @Test
    fun `buildDisplayText returns original text for content without directives`() {
        val result = DirectiveSegmenter.buildDisplayText("Hello world", TEST_LINE_ID, emptyMap())
        assertEquals("Hello world", result.displayText)
        assertTrue(result.directiveDisplayRanges.isEmpty())
    }

    @Test
    fun `buildDisplayText replaces directive with result`() {
        val content = "[42]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("42", result.displayText)
        assertEquals(1, result.directiveDisplayRanges.size)

        val range = result.directiveDisplayRanges[0]
        assertEquals(0..3, range.sourceRange)
        assertEquals(0..1, range.displayRange)
        assertEquals("[42]", range.sourceText)
        assertEquals("42", range.displayText)
        assertTrue(range.isComputed)
        assertFalse(range.hasError)
    }

    @Test
    fun `buildDisplayText handles mixed content`() {
        val content = "Hello [42] world"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        // "[42]" (4 chars) replaced with "42" (2 chars)
        assertEquals("Hello 42 world", result.displayText)
        assertEquals(1, result.directiveDisplayRanges.size)

        val range = result.directiveDisplayRanges[0]
        assertEquals(6..9, range.sourceRange)  // "[42]" in original
        assertEquals(6..7, range.displayRange) // "42" in display
    }

    @Test
    fun `buildDisplayText handles uncomputed directive`() {
        val content = "[42]"
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, emptyMap())

        // Should show source text when not computed
        assertEquals("[42]", result.displayText)
        assertEquals(1, result.directiveDisplayRanges.size)
        assertFalse(result.directiveDisplayRanges[0].isComputed)
    }

    @Test
    fun `buildDisplayText handles string directive`() {
        val content = "[\"hello\"]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("hello", result.displayText)
    }

    @Test
    fun `buildDisplayText handles multiple directives with different lengths`() {
        val content = "[100] [200]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("100 200", result.displayText)
        assertEquals(2, result.directiveDisplayRanges.size)

        // First directive
        val range1 = result.directiveDisplayRanges[0]
        assertEquals("[100]", range1.sourceText)
        assertEquals("100", range1.displayText)
        assertEquals(0..2, range1.displayRange)

        // Second directive
        val range2 = result.directiveDisplayRanges[1]
        assertEquals("[200]", range2.sourceText)
        assertEquals("200", range2.displayText)
        assertEquals(4..6, range2.displayRange)
    }

    // endregion

    // region alarm directive display range tests

    @Test
    fun `buildDisplayText sets isAlarm true for alarm directive`() {
        val content = "[alarm(\"abc123\")]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals(1, result.directiveDisplayRanges.size)
        val range = result.directiveDisplayRanges[0]
        assertTrue(range.isAlarm)
        assertEquals("abc123", range.alarmId)
        assertFalse(range.isView)
        assertFalse(range.isButton)
    }

    @Test
    fun `buildDisplayText alarm directive displays as clock emoji`() {
        val content = "[alarm(\"test\")]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("⏰", result.displayText)
        assertEquals("⏰", result.directiveDisplayRanges[0].displayText)
    }

    @Test
    fun `buildDisplayText alarm with surrounding text`() {
        val content = "Buy milk [alarm(\"x\")] today"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("Buy milk ⏰ today", result.displayText)
        val range = result.directiveDisplayRanges[0]
        assertTrue(range.isAlarm)
        assertEquals("x", range.alarmId)
    }

    @Test
    fun `buildDisplayText non-alarm directive has isAlarm false`() {
        val content = "[42]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals(1, result.directiveDisplayRanges.size)
        assertFalse(result.directiveDisplayRanges[0].isAlarm)
        assertNull(result.directiveDisplayRanges[0].alarmId)
    }

    @Test
    fun `buildDisplayText alarm sourceRange covers entire directive text`() {
        // Tap handlers use `charOffset in sourceRange` to detect alarm taps,
        // so sourceRange must span the full [alarm("id")] text
        val content = "Buy milk [alarm(\"abc\")] today"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        val range = result.directiveDisplayRanges[0]
        val directiveText = content.substring(range.sourceRange.first, range.sourceRange.last + 1)
        assertEquals("[alarm(\"abc\")]", directiveText)
        // Every character offset within the directive source range should match
        for (offset in range.sourceRange) {
            assertTrue("offset $offset should be in sourceRange", offset in range.sourceRange)
        }
        // Offsets just outside should NOT match
        assertFalse(range.sourceRange.first - 1 in range.sourceRange)
        assertFalse(range.sourceRange.last + 1 in range.sourceRange)
    }

    @Test
    fun `buildDisplayText alarm displayRange is single emoji width`() {
        // Display range for ⏰ must be exactly 1 char wide.
        // Tap handlers extend by +1 for cursor-position semantics.
        val content = "Task [alarm(\"x\")]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        val range = result.directiveDisplayRanges[0]
        assertEquals("⏰", range.displayText)
        // Display range length = 1 (single emoji character)
        assertEquals(1, range.displayRange.last - range.displayRange.first + 1)
    }

    @Test
    fun `buildDisplayText alarm with prefix adjustment has correct ranges`() {
        // End-to-end: checkbox prefix "☐ " (2 chars) + content with alarm
        val content = "[alarm(\"z\")]"
        val key = DirectiveResult.hashDirective("[alarm(\"z\")]")
        val results = mapOf(key to DirectiveFinder.executeDirective("[alarm(\"z\")]").result)

        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)

        assertEquals("⏰", result.displayText)
        val range = result.directiveDisplayRanges[0]
        assertTrue(range.isAlarm)
        assertEquals("z", range.alarmId)
        // Source range should cover the content-only text [alarm("z")]
        assertEquals(0..11, range.sourceRange)
        // Display range should be just the emoji
        assertEquals(0..0, range.displayRange)
    }

    // endregion

    // region adjustKeysForPrefix tests

    @Test
    fun `adjustKeysForPrefix returns same map when prefixLength is 0`() {
        val results = mapOf(
            "test-line:5" to DirectiveResult.success(NumberVal(42.0))
        )
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, TEST_LINE_ID, 0)
        assertSame(results, adjusted)
    }

    @Test
    fun `adjustKeysForPrefix subtracts prefix length from offset`() {
        val result = DirectiveResult.success(NumberVal(42.0))
        val results = mapOf("test-line:15" to result)
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, TEST_LINE_ID, 2)
        assertTrue(adjusted.containsKey("test-line:13"))
        assertFalse(adjusted.containsKey("test-line:15"))
        assertEquals(result, adjusted["test-line:13"])
    }

    @Test
    fun `adjustKeysForPrefix only adjusts keys matching lineId`() {
        val r1 = DirectiveResult.success(NumberVal(1.0))
        val r2 = DirectiveResult.success(NumberVal(2.0))
        val results = mapOf("test-line:10" to r1, "other-line:10" to r2)
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, TEST_LINE_ID, 2)
        // test-line key adjusted
        assertTrue(adjusted.containsKey("test-line:8"))
        // other-line key unchanged
        assertTrue(adjusted.containsKey("other-line:10"))
    }

    @Test
    fun `adjustKeysForPrefix handles multiple directives on same line`() {
        val r1 = DirectiveResult.success(NumberVal(1.0))
        val r2 = DirectiveResult.success(NumberVal(2.0))
        val results = mapOf("test-line:5" to r1, "test-line:20" to r2)
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, TEST_LINE_ID, 3)
        assertTrue(adjusted.containsKey("test-line:2"))
        assertTrue(adjusted.containsKey("test-line:17"))
    }

    @Test
    fun `adjustKeysForPrefix skips keys where adjusted offset would be negative`() {
        val result = DirectiveResult.success(NumberVal(42.0))
        val results = mapOf("test-line:1" to result)
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, TEST_LINE_ID, 5)
        // Offset 1 - 5 = -4, should keep original key
        assertTrue(adjusted.containsKey("test-line:1"))
    }

    @Test
    fun `adjustKeysForPrefix works with noteId-based keys`() {
        val result = DirectiveResult.success(NumberVal(42.0))
        val noteId = "abc123"
        val results = mapOf("$noteId:15" to result)
        val adjusted = DirectiveSegmenter.adjustKeysForPrefix(results, noteId, 3)
        assertTrue(adjusted.containsKey("$noteId:12"))
        assertFalse(adjusted.containsKey("$noteId:15"))
    }

    // endregion

    // region prefix key adjustment integration tests (end-to-end with buildDisplayText)

    @Test
    fun `buildDisplayText finds result when keys adjusted for checkbox prefix`() {
        // Simulate: full line is "☐ [alarm("x")]" (prefix "☐ " = 2 chars)
        // parseAllDirectiveLocations would find directive at offset 2 in full text
        // But buildDisplayText receives content "[alarm("x")]" and looks up at offset 0
        val content = "[alarm(\"x\")]"
        val key = DirectiveResult.hashDirective("[alarm(\"x\")]")
        val results = mapOf(key to DirectiveFinder.executeDirective("[alarm(\"x\")]").result)

        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)
        assertEquals("⏰", result.displayText)
        assertTrue(result.directiveDisplayRanges[0].isAlarm)
    }

    @Test
    fun `buildDisplayText finds result when keys adjusted for bullet prefix`() {
        val content = "Buy milk [alarm(\"y\")]"
        val lineId = "line-5"
        val key = DirectiveResult.hashDirective("[alarm(\"y\")]")
        val results = mapOf(key to DirectiveFinder.executeDirective("[alarm(\"y\")]").result)

        val result = DirectiveSegmenter.buildDisplayText(content, lineId, results)
        assertEquals("Buy milk ⏰", result.displayText)
    }

    @Test
    fun `buildDisplayText finds result when keys adjusted for tab prefix`() {
        val content = "[42]"
        val key = DirectiveResult.hashDirective("[42]")
        val results = mapOf(key to DirectiveFinder.executeDirective("[42]").result)

        val result = DirectiveSegmenter.buildDisplayText(content, TEST_LINE_ID, results)
        assertEquals("42", result.displayText)
    }

    @Test
    fun `buildDisplayText finds result when keys adjusted for tab plus checkbox prefix`() {
        // "\t☐ [alarm("z")]" — prefix is "\t☐ " = 3 chars
        val content = "[alarm(\"z\")]"
        val noteId = "testNote"
        val key = DirectiveResult.hashDirective("[alarm(\"z\")]")
        val results = mapOf(key to DirectiveFinder.executeDirective("[alarm(\"z\")]").result)

        val result = DirectiveSegmenter.buildDisplayText(content, noteId, results)
        assertEquals("⏰", result.displayText)
        assertTrue(result.directiveDisplayRanges[0].isAlarm)
        assertEquals("z", result.directiveDisplayRanges[0].alarmId)
    }

    // endregion

    // region executeAllDirectives key format tests

    @Test
    fun `executeAllDirectives produces position-independent keys`() {
        // Two identical directives at different positions produce the same key
        val content = "[42] and [42]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        // Both [42] directives share the same hash key, so only 1 entry
        assertEquals(1, results.size)
        val key = results.keys.first()
        assertEquals(DirectiveResult.hashDirective("[42]"), key)
    }

    @Test
    fun `executeAllDirectives produces different keys for different directives`() {
        val content = "[42] and [43]"
        val results = DirectiveFinder.executeAllDirectives(content, TEST_LINE_ID)
        assertEquals(2, results.size)
        assertTrue(results.containsKey(DirectiveResult.hashDirective("[42]")))
        assertTrue(results.containsKey(DirectiveResult.hashDirective("[43]")))
    }

    // endregion
}
